src/Serilog.Sinks.MSSqlServer/Sinks/MSSqlServer/MSSqlServerSinkTraits.cs

```diff
@@ -0,0 +1,330 @@
+        public string TableName { get; }
+        public string SchemaName { get; }
+        public ColumnOptions ColumnOptions { get; }
+        public IFormatProvider FormatProvider { get; }
+        public JsonFormatter JsonFormatter { get; }
+        public ISet<string> AdditionalDataColumnNames { get; }
+        public DataTable EventTable { get; }
+
+        public MSSqlServerSinkTraits(string connectionString, string tableName, string schemaName, ColumnOptions columnOptions, IFormatProvider formatProvider, bool autoCreateSqlTable)
+        {
+            if (string.IsNullOrWhiteSpace(connectionString))
+                throw new ArgumentNullException(nameof(connectionString));
+
+            if (string.IsNullOrWhiteSpace(tableName))
+                throw new ArgumentNullException(nameof(tableName));
+
+            ConnectionString = connectionString;
+            TableName = tableName;
+            SchemaName = schemaName;
+            ColumnOptions = columnOptions ?? new ColumnOptions();
+            FormatProvider = formatProvider;
+
+            if (ColumnOptions.AdditionalDataColumns != null)
+                AdditionalDataColumnNames = new HashSet<string>(ColumnOptions.AdditionalDataColumns.Select(c => c.ColumnName), StringComparer.OrdinalIgnoreCase);
+
+            if (ColumnOptions.Store.Contains(StandardColumn.LogEvent))
+                JsonFormatter = new JsonFormatter(formatProvider: formatProvider);
+
+            EventTable = CreateDataTable();
+
+            if (autoCreateSqlTable)
+            {
+                try
+                {
+                    SqlTableCreator tableCreator = new SqlTableCreator(connectionString, SchemaName);
+                    tableCreator.CreateTable(EventTable);
+                }
+                catch (Exception ex)
+                {
+                    SelfLog.WriteLine("Exception {0} caught while creating table {1} to the database specified in the Connection string.", ex, tableName);
+                }
+
+            }
+        }
+
+        /// <summary>Gets a list of the column names paired with their values to emit for the specified <paramref name="logEvent"/>.</summary>
+        /// <param name="logEvent">The log event to emit.</param>
+        /// <returns>
+        /// A list of mappings between column names and values to emit to the database for the specified <paramref name="logEvent"/>.
+        /// </returns>
+        public IEnumerable<KeyValuePair<string, object>> GetColumnsAndValues(LogEvent logEvent)
+        {
+            foreach (var column in ColumnOptions.Store)
+            {
+                yield return GetStandardColumnNameAndValue(column, logEvent);
+            }
+
+            if (ColumnOptions.AdditionalDataColumns != null)
+            {
+                foreach (var columnValuePair in ConvertPropertiesToColumn(logEvent.Properties))
+                    yield return columnValuePair;
+            }
+        }
+
+        public void Dispose()
```

Comment:
Maybe implement Dispose pattern here as well.

Reply:
This class is sealed.